metis-core/metis-core-common/src/main/java/eu/europeana/metis/core/workflow/plugins/AbstractIndexPlugin.java

```diff
@@ -0,0 +1,77 @@
+package eu.europeana.metis.core.workflow.plugins;
+
+import eu.europeana.cloud.service.dps.DpsTask;
+import eu.europeana.cloud.service.dps.metis.indexing.TargetIndexingDatabase;
+import eu.europeana.cloud.service.dps.metis.indexing.TargetIndexingEnvironment;
+
+/**
+ * Represents an index plugin.
+ *
+ * @param <M> the type of the plugin metadata in the plugin
+ */
+public abstract class AbstractIndexPlugin<M extends AbstractIndexPluginMetadata> extends AbstractExecutablePlugin<M> {
+
+  protected final String topologyName = Topology.INDEX.getTopologyName();
+  // TODO: 01/11/2021 The correct values should be updated with a script for the latest preview and publish executions, during release
+  //Initialize with negative for past executions
+  private int totalDatabaseRecords = -1;
+
+  /**
+   * Required by (de)serialization in db.
+   * <p>It is not to be used manually</p>
+   */
+  AbstractIndexPlugin() {
+    //Required by (de)serialization in db
+  }
+
+  /**
+   * Constructor to initialize the plugin with pluginMetadata required and the pluginType.
+   *
+   * @param pluginType a {@link PluginType} related to the implemented plugin
+   * @param pluginMetadata the plugin metadata
+   */
+  public AbstractIndexPlugin(PluginType pluginType, M pluginMetadata) {
+    super(pluginType, pluginMetadata);
+  }
+
+  @Override
+  public String getTopologyName() {
+    return topologyName;
+  }
+
+  @Override
+  public DpsTask prepareDpsTask(String datasetId, EcloudBasePluginParameters ecloudBasePluginParameters) {
+    return createDpsTaskForIndexPlugin(ecloudBasePluginParameters, datasetId,
+        getPluginMetadata().isIncrementalIndexing(),
+        getPluginMetadata().getHarvestDate(),
+        getPluginMetadata().isUseAlternativeIndexingEnvironment(),
+        getPluginMetadata().isPreserveTimestamps(),
+        getPluginMetadata().getDatasetIdsToRedirectFrom(),
+        getPluginMetadata().isPerformRedirects(), getTargetIndexingDatabase().name());
```

Comment:
I don't know where the method `createDpsTaskForIndexPlugin` comes from, but I see a lot of the input variables come from the `getPluginMetadata`, wouldn't be easier to just use a `pluginMetadata` as one of the inputs?

Reply:
The `createDpsTaskForIndexPlugin` comes from the `AbstractExecutablePlugin` which is higher in the hierarchy.
It makes sense what you are asking so I gave it some thought before replying.
Using a `pluginMetadata` is possible as long as it's a specific type. So to pass it further we'd need to cast it to the right type either when calling the method or inside it.
Where it is now, it's basically transparent and automatically recognized.
Also the `createDpsTaskForIndexPlugin` is there for other potential uses(even though we are not using it) without having an explicit plugin required.
What I don't like is that the higher class need to explicitly know the below class..
What we could do though is something similar to `TransformationPlugin`.
Having said that, we might need more time to think about it, since we should have a look at the other plugins at the same time.
So I'll leave it as is for now.

Reply:
I see your point, makes sense for the higher class to not know explicitly the below class. 
The approach done in `TransformationPlugin` seems to be the more feasible solution for this.
Perhaps we could create a Technical Debt ticket to tackle this?

Reply:
Sure, we can work with investigative td ticket.